### PR TITLE
feat: custom gas token semantics

### PR DESCRIPTION
**Pull request by tynes**
**Description**

Clean up the semantics of the custom gas token feature

Corresponding specs: https://github.com/ethereum-optimism/specs/pull/169

### CODE REVIEW

File: packages/contracts-bedrock/src/L2/L2StandardBridge.sol [hunk 1]
```diff
@@ -100,6 +100,7 @@ contract L2StandardBridge is StandardBridge, ISemver {
         virtual
         onlyEOA
     {
+        require(isCustomGasToken() == false, "L2StandardBridge: not supported with custom gas token");
```
Comment (mds1) [inline packages/contracts-bedrock/src/L2/L2StandardBridge.sol:104]: Should we also mention this is going to be deprecated in the future, as in specs?
```suggestion
        // These methods are deprecated and subject to be removed in the future.
        require(isCustomGasToken() == false, "L2StandardBridge: not supported with custom gas token");
```

```suggestion
        require(isCustomGasToken() == false, "L2StandardBridge: not supported with custom gas token, and may be removed in the future");
```

File: packages/contracts-bedrock/test/L2/L2StandardBridge.t.sol
Comment (mds1) [inline packages/contracts-bedrock/test/L2/L2StandardBridge.t.sol:1]: Code LGTM, just trying to make sure we have tests for all the revert cases. Our new spec, annotated with the test cases for the reverting cases, is below. I put a check emoji for tests I see and an X emoji for tests that I think are missing. (feel free to adjust naming convention)

When custom gas token is not used:
1. zero value + ETH + no custom gas token → _initiateBridgeETH → ok
2. nonzero value + ETH + no custom gas token → _initiateBridgeETH → ok
3. zero value + ERC20 + no custom gas token  → _initiateBridgeERC20 → ok
4. nonzero value + ERC20 + no custom gas token  → _initiateBridgeERC20 → revert
    1. ✅ `test_withdraw_erc20WithValue_reverts`
    2. ✅ `test_withdrawTo_erc20WithValue_reverts` 

When custom gas token is used:
1. zero value + ETH + custom gas token → _initiateBridgeETH → revert
    1. ✅ `test_withdraw_customGasToken_reverts`
    3. ✅ `test_withdrawTo_customGasToken_reverts`
2. nonzero value + ETH + custom gas token → _initiateBridgeETH → revert
    1. ✅ `test_withdraw_customGasTokenWithValue_reverts`
    2. ✅ `test_withdrawTo_customGasTokenWithValue_reverts`
4. zero value + ERC20 + custom gas token  → _initiateBridgeERC20 → revert
    2. ✅ `test_withdrawERC20_customGasToken_reverts`
    1. ✅ `test_withdrawToERC20_customGasToken_reverts`  
5. nonzero value + ERC20 + custom gas token  → _initiateBridgeERC20 → revert
    1. ✅  `test_withdrawERC20WithValue_customGasToken_reverts`
    1. ✅  `test_withdrawToERC20_customGasToken_reverts`  

And for `finalizeDeposit`
1. ✅ `test_finalizeDeposit_depositingETH_customGasToken_reverts`
1. ✅ `test_finalizeDeposit_depositingERC20_customGasToken_reverts`

Comment (tynes) [reply packages/contracts-bedrock/test/L2/L2StandardBridge.t.sol:1]: I have addressed this in https://github.com/ethereum-optimism/optimism/pull/10374/commits/f48de83bfd50939f54cc4084c0eac402422d8b60

File: packages/contracts-bedrock/src/L2/L2StandardBridge.sol (hunk 1 above)
Comment (tynes) [reply packages/contracts-bedrock/src/L2/L2StandardBridge.sol:104]: I added these comments to the natspec

File: packages/contracts-bedrock/test/L2/L2StandardBridge.t.sol
Comment (mds1) [reply packages/contracts-bedrock/test/L2/L2StandardBridge.t.sol:1]: Perfect. I've updated my comment to with the new test names you've used. Thank you!

**Review by mds1 — APPROVED** [review]

**Pull request merged**